### PR TITLE
HDDS-6553. Incorrect timeout when checking balancer iteration result.

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

The scenario is as follows:

1. ContainerBalancer trying to move the following containers:
    1. container 1, took 20 minutes
    1. container 2, took 40 minutes
    1. container 3, took 60 minutes
2. When checking if a move finishes, the timeout is set to 30minutes.
3. When ContainerBalancer waited 20 minutes, it records container1 as finished.
4. Then Containerbalancer will wait another 30 minutes for container 2. Although container 2 took 40 minutes, ContainerBalancer will still record container 2 as a finished move.
5. The same issue happens when checking move result for container 3.

This ticket is to fix this issue.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6553

## How was this patch tested?

unit test.

### CODE REVIEW

**Comment by symious** [comment]
@lokeshj1703 @siddhantsangwan Could you help to review this PR?

**Review by siddhantsangwan — COMMENTED** [review]
@symious This is a nice improvement. I have a few comments.

**Comment by symious** [comment]
@siddhantsangwan Thanks for the review. Updated the patch, please have a check.

**Review by lokeshj1703 — COMMENTED** [review]
Thanks for working on this @symious! Good catch.
The changes look good to me. I have a few comments inline.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 1]
```diff
@@ -462,49 +462,80 @@ private void checkIterationResults(boolean isMoveGenerated,
...
+                ex.getCause());
           }
         } else {
-          if (LOG.isDebugEnabled()) {
-            LOG.debug("Container move for container {} to target {} failed: {}",
-                moveSelection.getContainerID(),
-                moveSelection.getTargetNode().getUuidString(), result);
+          ReplicationManager.MoveResult result = future.join();
+          if (result == ReplicationManager.MoveResult.COMPLETED) {
+            try {
```
[the remark this answers is not in the document]
Comment (symious) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:502]: @lokeshj1703 Thanks for the review.
There is a problem if added the post process of futures in whenComplete. The moves after timeout can not be cancelled, so the move will keep running until finished.
As the TestContainerBalancer.checkIterationResultTimeout shows in https://github.com/symious/ozone/tree/HDDS-6553-
2, the metrics of numContainerMovesInLatestIteration will always be larger than 1, which are incremented by the timeouted futures.

[the remark this answers is not in the document]
Comment (symious) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:501]: Updated the patch for timout moves metrics and log exceptions, please have a check.

[the remark this answers is not in the document]
Comment (lokeshj1703) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:501]: I think we can just run a for loop and cancel the futures(It only cancels the futures which are not already completed, I think). WhenComplete would make the implementation simpler to understand.

[the remark this answers is not in the document]
Comment (symious) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:501]: Understand that using whenComplete would make the code more succinct.

But according to https://stackoverflow.com/a/65096960, seems completableFuture is not good to handle cancel. 
The current for loop process is referenced from https://stackoverflow.com/a/65221108.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 2]
```diff
@@ -462,49 +462,80 @@ private void checkIterationResults(boolean isMoveGenerated,
...
+        moveSelectionToFutureMap.values()
+            .toArray(new CompletableFuture[moveSelectionToFutureMap.size()]));
+    try {
+      allFuturesResult.get(config.getMoveTimeout().toMillis(),
+          TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (Exception e) {
+      // For exceptions other than InterruptedException, we ignore them.
+      e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (lokeshj1703) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:485]: This still needs to be handled in the latest commit.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 3]
```diff
@@ -603,25 +585,46 @@ private boolean checkIterationLimits() {
...
     CompletableFuture<ReplicationManager.MoveResult> future;
     try {
+      ContainerInfo containerInfo = containerManager.getContainer(containerID);
       future = replicationManager
-          .move(container, source, moveSelection.getTargetNode());
+          .move(containerID, source, moveSelection.getTargetNode())
+          .whenComplete((result, ex) -> {
+            if (ex != null) {
+              LOG.info("Container move for container {} from source {} to " +
+                      "target {} failed with exceptions",
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:597]: missing {} after exceptions

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 4]
```diff
@@ -603,25 +585,46 @@ private boolean checkIterationLimits() {
...
-      LOG.warn("Could not find Container {} for container move", container, e);
+      LOG.warn("Could not find Container {} for container move",
+          containerID, e);
       return false;
     } catch (NodeNotFoundException e) {
-      LOG.warn("Container move failed for container {}", container, e);
+      LOG.warn("Container move failed for container {}", containerID, e);
       return false;
     }
     if (future.isDone()) {
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:626]: Let's add an empty line before this.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 5]
```diff
@@ -603,25 +585,46 @@ private boolean checkIterationLimits() {
...
+                      "target {} failed with exceptions",
+                  containerID.toString(),
+                  source.getUuidString(),
+                  moveSelection.getTargetNode().getUuidString(), ex);
+            } else {
+              if (result == ReplicationManager.MoveResult.COMPLETED) {
+                metrics.incrementDataSizeMovedGBInLatestIteration(
+                    containerInfo.getUsedBytes() / OzoneConsts.GB);
+                metrics.incrementNumContainerMovesInLatestIteration(1);
+                LOG.info(
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:606]: We can make this debug and wrap it inside LOG.isDebugEnabled check.
```
if (LOG.isDebugEnabled) {
   LOG.info...
}
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 6]
```diff
@@ -603,25 +585,46 @@ private boolean checkIterationLimits() {
...
+                    "Container move completed for container {} to target {}",
+                    containerID,
+                    moveSelection.getTargetNode().getUuidString());
+              } else {
+                LOG.debug(
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:611]: We can make this info or warn.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerMetrics.java [hunk 7]
```diff
@@ -44,6 +44,10 @@ public final class ContainerBalancerMetrics {
       "in the latest iteration.")
   private MutableCounterLong numContainerMovesInLatestIteration;
 
+  @Metric(about = "Number of timeout container moves performed by " +
+      "Container Balancer in the latest iteration.")
+  private MutableCounterLong numTimeoutContainerMovesInLatestIteration;
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerMetrics.java:49]: Can we rename this to `numContainerMovesTimeoutInLatestIteration` and `numContainerMovesCompletedInLatestIteration`?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancer.java [hunk 8]
```diff
@@ -642,6 +642,37 @@ public void checkIterationResult()
...
+    sleepWhileBalancing(2000);
+
+    /*
+    According to the setup and configurations, this iteration's result should
+    be ITERATION_COMPLETED.
+     */
+    Assert.assertEquals(ContainerBalancer.IterationResult.ITERATION_COMPLETED,
+        containerBalancer.getIterationResult());
+    Assert.assertEquals(1,
+        containerBalancer.getMetrics().getNumContainerMovesInLatestIteration());
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancer.java:671]: Can we also add assertion for timed out metrics?

**Review by lokeshj1703 — COMMENTED** [review]
@symious Thanks for updating the PR! The changes look good to me.
I have a few minor comments inline.

**Comment by symious** [comment]
@lokeshj1703 Thanks for the review.
Fixed all the comments exception the exception one. The exception can be logged without a explicit "{}".

**Review by lokeshj1703 — APPROVED** [review]
@symious Thanks for updating the PR! The changes look good to me. +1.

**Comment by lokeshj1703** [comment]
@symious Thanks for the contribution! @siddhantsangwan Thanks for the review! I have committed the PR to master branch.

**Comment by symious** [comment]
@lokeshj1703 @siddhantsangwan Thanks for the review.

**Pull request merged**